### ISSUE TITLE
Picocli extension - consider generating command metadata during build

### ISSUE THREAD

**Issue by mkouba** [quarkusio/quarkus#16355]
ATM picocli builds the `CommandSpec` metadata using reflection every time the application is started. This is OK in most situations but may represent considerable overhead for applications that finish instantly, e.g. some Quarkus CLI commands. It might make sense to analyze the commands during build and either generate some "builder" classes or record the metadata (e.g. using `AnnotationProxyBuildItem`) and build the metadata in a recorder method.

**Comment by mkouba** [comment]
WDYT @mgorniew?

**Comment by gsmet** [comment]
Yeah I had the same thought the other day. I'm just unsure how much time it takes. Given displaying the help is slow enough even in native, I suppose it takes quite some time.

**Comment by mkouba** [comment]
> I'm just unsure how much time it takes.

My super unreliable measurement showed that it could be hundreds of ms (quarkus cli has 1 top command + 10 subcommand classes).

**Comment by maxandersen** [comment]
> Given displaying the help is slow enough even in native, I suppose it takes quite some time.

big part of this one is probably due the console width calculation which can be expensive even in native. 

that said - i do think picoclis annotation parsing is quite slow - its signficiant in jbang too.

**Comment by mkouba** [comment]
> > Given displaying the help is slow enough even in native, I suppose it takes quite some time.
> 
> big part of this one is probably due the console width calculation which can be expensive even in native.

This part should be solved in https://github.com/quarkusio/quarkus/pull/16320.

**Comment by remkop** [comment]
I’ve been wanting to do something like this: https://github.com/remkop/picocli/issues/539 but never got around to it.

**Comment by remkop** [comment]
> Yeah I had the same thought the other day. I'm just unsure how much time it takes. Given displaying the help is slow enough even in native, I suppose it takes quite some time.

Can you put some numbers on that? How long does displaying the help take in native and in JIT mode?

**Comment by quintesse** [comment]
@remkop I don't have the exact numbers at hand (they're somewhere in a chat) but when I was investigating the slow startup of Jbang I found that around .6s of a 1.2s startup was due to Picocli setup. For a tool that's supposed to enable Java as a scripting language adding more than a second to the startup will make people think twice before dropping Perl/Node/Bash in favor of Java. As a comparison java and javac start up in about 0.1s and 0.3s so ideally we don't want to be adding too much on top of that.

**Comment by ebullient** [comment]
Now that I've done the crazy things and restructured the entire CLI to be a little cleaner.. I could also just say eff the annotations and code it directly.

Annotations are cool and all, but we don't actually have to use them..

**Comment by quintesse** [comment]
@ebullient I was considering the same for Jbang and my worry is that it will become less maintainable (somewhat ) and less accessible to people who are not familiar with the code.

**Comment by maxandersen** [comment]
Annotations makes sense to be declarative and quarkus be smarter about it. 

Even doing it programmatically will have an overhead compared to have the quarkus picocli extension do it during build and just load static model at startup.

**Comment by maxandersen** [comment]
@quintesse consider converting https://github.com/quintesse/jbang/commit/15e7e85a4932722353a95483787271da083d50a6 to quarkus extension?

**Comment by lppedd** [comment]
Is this still valid?
I've built native executables on Windows, and I've noticed the Micronaut starter project measures at around 200-250ms for `-h`, while the Quarkus one at around 350ms+. Is this to be expected?
It's not a big deal, but it's noticeable.

**Comment by maxandersen** [comment]
@lppedd Is micronaut using picocli or something else? I haven't made the comparison. 

But yes, with picocli i'm convinced you'll be able to shave of 100-200 ms of startup if the picocli extension would do its annotation model scanning at build time rather than at runtime.

**Comment by ebullient** [comment]
That only works of that marked class doesn't use any beans. Curious if using a static init to create/initialize the command instance programmatically would make a difference (to prove your point), as static init is baked into the native image, IIRC.

**Comment by quintesse** [comment]
@ebullient that's more or less what I did here: https://github.com/quintesse/jbang/blob/15e7e85a4932722353a95483787271da083d50a6/annotation/src/main/java/MyProcessor.java , it's a build processor that takes that PicoCli annotations at build time and generates code that does the same thing.

Unfortunately there was (is?) no API in PicoCli that I could hook into to turn annotations into their appropriate models so I had to do it the hard way (which means there's only partial support for PicoCli's features, basically just enough to make it work with our project).

But perhaps someone can use it to show if doing it like that makes a difference or not.

**Comment by quintesse** [comment]
> Am I correct?

Correct. And I created the processor to see if I could create the model at compile time and I because we want our application to be as small as possible I didn't want to depend on any framework, that's why it was not written with Quarkus in mind. I abandoned the PoC when I found no noticeable improvement in startup time.

But the experiment's results at least suggest it's not the annotation parsing that's slowing things down but constructing / initializing the model.

So that's why it was suggested that perhaps Quarkus' way of statically initializing objects could be useful.

(I'm probably repeating myself too much, but I just want to make sure things are clear heh)

**Comment by mkouba** [comment]
Thanks @quintesse I'll try to do some quick tests (when I have spare cycles ;-).
